dae/dae/annotation/annotate_vcf.py

```diff
@@ -27,78 +34,150 @@ def configure_argument_parser() -> argparse.ArgumentParser:
-    )
-    header.add_meta(
-        "pipeline_annotation_tool", f"{' '.join(sys.argv)}")
-    for attribute in annotation_attributes:
+def update_header(variant_file, pipeline):
+    """Update a variant file's header with annotation pipeline scores."""
+    header = variant_file.header
+    header.add_meta("pipeline_annotation_tool", "GPF variant annotation.")
+    header.add_meta("pipeline_annotation_tool", f"{' '.join(sys.argv)}")
+    for attribute in pipeline.annotation_schema.names:
         description = pipeline.annotation_schema[attribute].description
         description = description.replace("\n", " ")
-        header.info.add(
-            attribute, "A", "String",
-            description)
+        header.info.add(attribute, "A", "String", description)
 
-    print(str(header), file=out_file, end="")
 
-    # handling the variants
-    with pipeline.open() as pipeline:
-        for vcf_var in in_file:
-            buffers: List[List] = [list([]) for _ in annotation_attributes]
+def annotate(input_file, region, pipeline, out_filepath):
+    """Annotate a region from a given input VCF file using a pipeline."""
+    in_file = VariantFile(input_file)
+    update_header(in_file, pipeline)
+    out_file = VariantFile(out_filepath, "w", header=in_file.header)
+    annotation_attributes = pipeline.annotation_schema.names
+    with pipeline.open():
+        for vcf_var in in_file.fetch(*region):
+            # pylint: disable=use-list-literal
+            buffers: List[List] = [list() for _ in annotation_attributes]
 
             if vcf_var.alts is None:
                 logger.info("vcf variant without alternatives: %s", vcf_var)
                 continue
 
             for alt in vcf_var.alts:
-                annotabale = VCFAllele(
-                    vcf_var.chrom, vcf_var.pos, vcf_var.ref, alt)
-                annotation = pipeline.annotate(annotabale)
+                annotation = pipeline.annotate(
+                    VCFAllele(vcf_var.chrom, vcf_var.pos, vcf_var.ref, alt)
+                )
                 for buff, attribute in zip(buffers, annotation_attributes):
-                    buff.append(str(annotation[attribute]))
+                    # TODO Ask what value to use for missing attr
+                    buff.append(str(annotation.get(attribute, "-")))
 
             for attribute, buff in zip(annotation_attributes, buffers):
-                vcf_var.info[attribute] = ",".join(buff)
-            print(str(vcf_var), file=out_file, end="")
-
+                vcf_var.info[attribute] = buff
+            out_file.write(vcf_var)
+    out_file.close()
     in_file.close()
 
-    if args.output != "-":
-        out_file.close()
+
+def combine(input_filepath, pipeline, part_filenames, output_filepath):
+    """Combine annotated region parts into a single VCF file."""
+    input_file = VariantFile(input_filepath)
+    update_header(input_file, pipeline)
+    with VariantFile(output_filepath, "w", header=input_file.header) as out:
```

Comment:
Include `input_file` to the context manager or use another `with` statement.

Do not forget to use https://docs.python.org/3/library/contextlib.html#contextlib.closing.